File: SDKs/Aspose_Cloud_SDK_For_Android/asposecloudsdk/src/main/java/com/aspose/cloud/sdk/storage/api/Folder.java
package com.aspose.cloud.sdk.storage.api;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Date;
import java.util.List;

import android.net.Uri;

import com.aspose.cloud.sdk.common.AsposeApp;
import com.aspose.cloud.sdk.common.BaseResponse;
import com.aspose.cloud.sdk.common.Utils;
import com.aspose.cloud.sdk.storage.model.DateDeserializer;
import com.aspose.cloud.sdk.storage.model.DiscUsageModel;
import com.aspose.cloud.sdk.storage.model.DiscUsageResponse;
import com.aspose.cloud.sdk.storage.model.FileExist;
import com.aspose.cloud.sdk.storage.model.FileExistResponse;
import com.aspose.cloud.sdk.storage.model.FileModel;
import com.aspose.cloud.sdk.storage.model.FolderResponse;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public class Folder {

	private static final String FOLDER_URI = AsposeApp.BASE_PRODUCT_URI + "/storage/folder/";
	private static final String FILE_URI = AsposeApp.BASE_PRODUCT_URI + "/storage/file/";
	private static final String EXIST_URI = AsposeApp.BASE_PRODUCT_URI + "/storage/exist/";
	private static final String DISC_URI = AsposeApp.BASE_PRODUCT_URI + "/storage/disc/";
	
	/**
	 * Get a list of all files and folders under the specified folder. Use empty string to specify root folder.
	 * @param folderPath Folder path
	 * @return Array of File Objects
	*/ 
	public static List<FileModel> getFilesList(String folderPath) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		List<FileModel> filesList = null;
			
		if(folderPath == null) {
			throw new IllegalArgumentException("Folder path cannot be null");
		}
		
		//build URL
      	String strURL = FOLDER_URI + Uri.encode(folderPath);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "GET");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
        GsonBuilder gsonBuilder = new GsonBuilder();
        gsonBuilder.registerTypeAdapter(Date.class, new DateDeserializer());
      	Gson gson = gsonBuilder.create();
      	FolderResponse folderResponse = gson.fromJson(responseJSONString, FolderResponse.class);
		if(folderResponse.getCode().equals("200") && folderResponse.getStatus().equals("OK")) {
			filesList = folderResponse.files;
		}
		
		return filesList;
	}

	/**
	 * Creates a folder under the specified path. If no path specified, creates a folder under the root folder.
	 * @param folderPath Folder path
	 * @return Boolean variable indicates whether folder created successfully
	*/ 
	public static boolean createFolder(String folderPath) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		boolean isFolderCreatedSuccessfully = false;
		
		if(folderPath == null) {
			throw new IllegalArgumentException("Folder path cannot be null");
		}
		
		//build URL
      	String strURL = FOLDER_URI + Uri.encode(folderPath);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "PUT");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFolderCreatedSuccessfully = true;
		}
		
		return isFolderCreatedSuccessfully;
	}
	
	/**
	 * Delete an empty folder from the storage. Use "FolderName/SubFolderName" for sub folders.
	 * @param folderPath Folder path
	 * @return Boolean variable indicates whether folder deleted successfully
	*/ 
	public static boolean deleteFolder(String folderPath) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		boolean isFolderDeletedSuccessfully = false;
		
		if(folderPath == null || folderPath.length() == 0) {
			throw new IllegalArgumentException("Folder path cannot be null or empty");
		}
		
		//build URL
      	String strURL = FOLDER_URI + Uri.encode(folderPath);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "DELETE");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFolderDeletedSuccessfully = true;
		}
		
		return isFolderDeletedSuccessfully;	
		
	}
	
	/**
	 * Delete a file from the storage. Use "FolderName/FileName" to specify a file under specific folder
	 * @param filePath File path
	 * @return Boolean variable indicates whether file deleted successfully
	*/ 
	public static boolean deleteFile(String filePath) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		boolean isFileDeletedSuccessfully = false;
		
		if(filePath == null || filePath.length() == 0) {
			throw new IllegalArgumentException("File path cannot be null or empty");
		}
		
		//build URL
      	String strURL = FILE_URI + Uri.encode(filePath);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "DELETE");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFileDeletedSuccessfully = true;
		}
		
		return isFileDeletedSuccessfully;	
	}

	/**
	 * Upload a file from your local machine to remote folder
	 * @param folderPath Folder path
	 * @param filePath File path on device
	 * @return Boolean variable that indicates whether file uploaded successfully
	*/ 
	public static boolean uploadFile(String localFilePath, String remoteFolderPath) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		Boolean isFileUploadedSuccessfully = false;
		
		if(localFilePath == null || localFilePath.length() == 0) {
			throw new IllegalArgumentException("Local file path cannot be null or empty");
		}
		
		if(remoteFolderPath == null) {
			throw new IllegalArgumentException("Remote folder path cannot be null or empty");
		}
		
		File localFile = new File(localFilePath);
		String fileName = localFile.getName();
		
		String remoteFilePath = remoteFolderPath.length() != 0? remoteFolderPath + "/" + fileName : fileName;
		
		//build URL
      	String strURL = FILE_URI + Uri.encode(remoteFilePath);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
		String responseJSONString = Utils.uploadFileBinary(localFile, signedURL, "PUT");

		//Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFileUploadedSuccessfully = true;
		}
		
		return isFileUploadedSuccessfully;
	}

	/**
	 * Checks whether file or folder exists on the Aspose storage
	 * @param fileName File name
	*/ 
	public static FileExist fileExist(String fileName) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		FileExist fileExist = null;
		
		if(fileName == null || fileName.length() == 0) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		//build URL
      	String strURL = EXIST_URI + Uri.encode(fileName);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "GET");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
      	Gson gson = new Gson();
      	FileExistResponse existResponse = gson.fromJson(responseJSONString, FileExistResponse.class);
		if(existResponse.getCode().equals("200") && existResponse.getStatus().equals("OK")) {
			fileExist = existResponse.fileExist;
		}
		
		return fileExist;
	}

	/**
	 * Provides the total and free disc size in bytes for your app
	*/ 
	public static DiscUsageModel getDiscUsage() throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		DiscUsageModel discUsage = null;
		
        //sign URL
        String signedURL = Utils.sign(DISC_URI);
        
        InputStream responseStream = Utils.processCommand(signedURL, "GET");
        String responseJSONString = Utils.streamToString(responseStream);
        
        //Parsing JSON
      	Gson gson = new Gson();
      	DiscUsageResponse existResponse = gson.fromJson(responseJSONString, DiscUsageResponse.class);
		if(existResponse.getCode().equals("200") && existResponse.getStatus().equals("OK")) {
			discUsage = existResponse.discUsage;
		}
		
		return discUsage;
	}

	/**
	 * Get file from Aspose server
	 * @param fileName File name
	*/ 
	public static InputStream getFile(String fileName) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		InputStream fileStream = null;
		
		if(fileName == null || fileName.length() == 0) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		//build URL
      	String strURL = FILE_URI + Uri.encode(fileName);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        fileStream = Utils.processCommand(signedURL, "GET");
        
        return fileStream;
	}
	
	/**
	 * Move file to designated location
	 * @param fileName File name
	 * @param srcFolderName Source folder name
	 * @param destFolderName Destination folder name
	 * @return Boolean variable that indicates whether file moved successfully to designated location
	*/ 
	public static boolean moveFile(String fileName, String srcFolderName, String destFolderName) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		boolean isFileMovedToAnotherLocationSuccessfully = false;
		
		if(fileName == null || fileName.length() == 0) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		if(destFolderName == null || destFolderName.length() == 0) {
			throw new IllegalArgumentException("Destination folder name cannot be null or empty");
		}
		
		String sourceFilePath;
		if(srcFolderName == null || srcFolderName.length() == 0) {
			sourceFilePath = Uri.encode(fileName);
		} else {
			sourceFilePath = Uri.encode(srcFolderName) + "/" + Uri.encode(fileName);
		}
		
		//build URL
      	String strURL = FILE_URI + sourceFilePath + "?dest=" + Uri.encode(destFolderName) + "/" + Uri.encode(fileName);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "POST");
        String responseJSONString = Utils.streamToString(responseStream);
		
        //Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFileMovedToAnotherLocationSuccessfully = true;
		}
        
		return isFileMovedToAnotherLocationSuccessfully;
	}
	
	/**
	 * Copy file to designated location
	 * @param fileName File name
	 * @param newDest Destination path
	 * @return Boolean variable that indicates whether file copied successfully to designated location
	*/ 
	public static boolean copyFile(String fileName, String newDest) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		boolean isFileCopiedToNewDestinationSuccessfully = false;
		
		if(fileName == null || fileName.length() == 0) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		if(newDest == null || newDest.length() == 0) {
			throw new IllegalArgumentException("newDest cannot be null or empty");
		}
		
		//build URL
      	String strURL = FILE_URI + Uri.encode(fileName) + "?newdest=" + Uri.encode(newDest) + "/" + Uri.encode(fileName);
        //sign URL
        String signedURL = Utils.sign(strURL);
        
        InputStream responseStream = Utils.processCommand(signedURL, "PUT");
        String responseJSONString = Utils.streamToString(responseStream);
		
        //Parsing JSON
      	Gson gson = new Gson();
      	BaseResponse baseResponse = gson.fromJson(responseJSONString, BaseResponse.class);
		if(baseResponse.getCode().equals("200") && baseResponse.getStatus().equals("OK")) {
			isFileCopiedToNewDestinationSuccessfully = true;
		}
		
		return isFileCopiedToNewDestinationSuccessfully;
	}
	
}

File: SDKs/Aspose_Cloud_SDK_For_Ruby/lib/Storage/folder.rb
module Aspose
  module Cloud
    module AsposeStorage
      # This class provides functionality to manage files in a Remote Aspose Folder
      class Folder

        def initialize
          @str_uri_folder = Aspose::Cloud::Common::Product.product_uri + '/storage/folder/'
          @str_uri_file = Aspose::Cloud::Common::Product.product_uri + '/storage/file/'
          @str_uri_exist = Aspose::Cloud::Common::Product.product_uri + '/storage/exist/'
          @str_uri_disc = Aspose::Cloud::Common::Product.product_uri + '/storage/disc/'
        end

=begin
  Uploads file from the local path to the remote folder
  @param string localFilePath represents full local file path and name.
=end        		 
        def upload_file(local_file, remote_folder='', storage_type='Aspose', storage_name='')
          raise 'Local file not specified' if local_file.empty?

          filename = File.basename(local_file)
          str_uri = "#{ Aspose::Cloud::Common::Product.product_uri }/storage/file/#{ remote_folder + '/' unless remote_folder.empty? }#{ filename }"
          str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

          signeduri = Aspose::Cloud::Common::Utils.sign(str_uri)
          response = JSON.parse(Aspose::Cloud::Common::Utils.upload_file_binary(local_file, signeduri))
          response['Status'].eql? 'OK'
        end

=begin
  Retrieves Files and Folder information from a remote folder
=end        
        def get_files(remote_folder_path='', storage_type='Aspose', storage_name='')
          str_uri = @str_uri_folder + remote_folder_path
          str_uri = str_uri[0..-2] if str_uri[-1].eql? '/'
          str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

          signed_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
          response = RestClient.get(signed_uri, :accept => 'application/json')

          JSON.parse(response)['Files']
        end

=begin
  Check if a file exists on the storage
  @param string filename Name of the file.
=end
        def file_exists(filename, storage_type = 'Aspose', storage_name = '')
            raise('Filename cannot be empty') if filename.empty?

            str_uri = @str_uri_exist + filename
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_str_uri = Aspose::Cloud::Common::Utils.sign(str_uri)

            response_stream = RestClient.get(signed_str_uri, {:accept => 'application/json'})
            JSON.parse(response_stream)['FileExist']['IsExist']
        end

=begin
  Delete a Particular File
  @param string filename Name of the file.
=end
        def delete_file(filename, storage_type = 'Aspose', storage_name = '')
            raise 'File name cannot be empty' if filename.empty?

            str_uri = @str_uri_file + filename
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_str_uri = Aspose::Cloud::Common::Utils.sign(str_uri)

            response_stream = RestClient.delete(signed_str_uri, {:accept => 'application/json'})
            JSON.parse(response_stream)['Code'].eql? 200
        end

=begin
  Create a New Folder
  @param string folder_name Name of the folder.
=end
        def create_folder (folder_name, storage_type = 'Aspose', storage_name='')
            raise 'Folder name cannot be empty' if folder_name.empty?
            str_uri = @str_uri_folder + folder_name
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
            response = RestClient.put(signed_uri, '', :accept => :json)
            JSON.parse(response)['Code'].eql? 200
        end

=begin
  Delete a Particular Folder
  @param string folder_name Name of the folder.
=end
        def delete_folder (folder_name, storage_type = 'Aspose', storage_name='')
            raise 'Folder name cannot be empty' if folder_name.empty?
            str_uri = @str_uri_folder + folder_name
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
            response = RestClient.delete(signed_uri, :accept => 'application/json')
            JSON.parse(response)['Code'].eql? 200
        end

=begin
  Get Disk Usage
=end
        def get_disc_usage (storage_type = 'Aspose', storage_name = '')
            str_uri = @str_uri_disc
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
            response = RestClient.get(signed_uri, :accept => 'application/json')
            JSON.parse(response)['DiscUsage']
        end

=begin
  Get file from storage
  @param string file_name Name of the file.
=end
        def get_file (file_name, storage_type = 'Aspose', storage_name = '')
            raise 'Filename cannot be empty' if file_name.empty?

            str_uri = @str_uri_file + file_name
            str_uri += append_storage(storage_name) unless storage_type.eql? 'Aspose'
            str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,'',storage_name,storage_type)

            signed_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
            RestClient.get(signed_uri, :accept => 'application/json')
        end
      end #Class Ends Here
    end
  end
end
